### PR TITLE
Custom tag & attributes to detect links

### PR DESCRIPTION
**Pull request by marapper**
Detect broken list in custom attributes.

For example if website uses lazyload:

```
   <img data-src="some-link.jpg" src="" alt="" class="lazyload" />
```

End-user can expand attributes list without forking of library:

```
await checker.check({
    path: 'http://example.com',
    linksAttributes: {
       'data-src': [
          'img'
       ]
    }
  });
```

### CODE REVIEW

**Comment by marapper** [comment]
And I noticed https://github.com/JustinBeckwith/linkinator/issues/70 

If I understand correctly crawler can also collect innerHTML of tags when find url in attributes. But what to do with self-cloed tags like img (collect alt-text?). And what to do with custom attributes on div for example - innerhtml in that cases can be as long as almost full body of response.

**Comment by JustinBeckwith** [comment]
This is super cool, thanks for the PR!  Initial thought: do we need the additional flag?  This feels like the kind of thing where we could just parse the `style` tags, and grab links in CSS (just like you've done), but just do it automatically.  Is there any danger in always doing it?

File: README.md [hunk 1]
```diff
@@ -128,6 +128,7 @@ Asynchronous method that runs a site wide scan. Options come in the form of an o
 - `port` (number) - When the `path` is provided as a local path on disk, the `port` on which to start the temporary web server.  Defaults to a random high range order port.
 - `recurse` (boolean) - By default, all scans are shallow.  Only the top level links on the requested page will be scanned.  By setting `recurse` to `true`, the crawler will follow all links on the page, and continue scanning links **on the same domain** for as long as it can go. Results are cached, so no worries about loops.
 - `linksToSkip` (array) - An array of regular expression strings that should be skipped during the scan.
+- `linksAttributes` (object) - Expand tag attributes to find URLs. Keys are attributes and values are array of tag to search.
```
Comment (JustinBeckwith) [inline README.md]: As mentioned, I think we could just drop this to be honest, and always parse `style`

File: src/links.ts [hunk 2]
```diff
@@ -45,6 +56,21 @@ export function getLinks(source: string, baseUrl: string): string[] {
 
 function parseAttr(name: string, value: string): string[] {
   switch (name) {
+    case 'style':
+      return value
```
Comment (JustinBeckwith) [inline src/links.ts]: There's a risk here of an empty `<style>` tag, no?

File: src/links.ts [hunk 3]
```diff
@@ -45,6 +56,21 @@ export function getLinks(source: string, baseUrl: string): string[] {
...
+        .split(';')
+        .filter(Boolean)
```
Comment (JustinBeckwith) [inline src/links.ts]: This feels a little magical.  Thinking it would more clear to do:
```js
.filter(line => !!line)
```

File: src/links.ts [hunk 4]
```diff
@@ -45,6 +56,21 @@ export function getLinks(source: string, baseUrl: string): string[] {
...
+        .map((rules: string) => {
+          const [attr, value] = rules.split(':');
```
Comment (JustinBeckwith) [inline src/links.ts]: I'm worried about this parser being a little naive.  For example:

```css
h4, h1 a:hover {
	text-decoration: none;
	color: #333;
}
```

Wouldn't the `:` in the `a:hover` cause some trouble in here?  I don't have an answer quite yet, but I'm wondering if we should look into a real CSS lexer/parser similar to cheerio.

File: README.md (hunk 1 above)
Comment (marapper) [reply README.md]: This is not for style its for custom `data-` attributes. There's no way to standatize it at all.

File: src/links.ts (hunk 2 above)
Comment (marapper) [reply src/links.ts]: Agreed.

File: src/links.ts (hunk 3 above)
Comment (marapper) [reply src/links.ts]: It's the same actually. But I fix it.

File: src/links.ts (hunk 4 above)
Comment (marapper) [reply src/links.ts]: Absolutely. Maybe PostCSS parser?

Comment (marapper) [reply src/links.ts]: Some note. This code is about style attribute not style tag so right here we have no problem with pseude-class & pseudo-elements (event with regexp).

Comment (marapper) [reply src/links.ts]: Feature is extracted to https://github.com/JustinBeckwith/linkinator/pull/113

**Comment by JustinBeckwith** [comment]
Thanks for the PR! But I don't think we're looking to add this feature right now.

**Pull request closed without merging**